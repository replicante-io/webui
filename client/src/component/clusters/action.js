'use strict';
//@flow

import type { ClusterMeta } from '../dashboard/action';


/** Name actions for type checking. */
export const CLUSTERS_LIST = 'CLUSTERS_LIST';
export const CLUSTERS_SEARCH = 'CLUSTERS_SEARCH';


/** Type enum of all possible fetch actions. */
export type ListClustersAction = {
  +type: typeof CLUSTERS_LIST,
  +clusters: Array<ClusterMeta>,
};

export type SearchClustersAction = {
  +type: typeof CLUSTERS_SEARCH,
  +search: string,
};

export type ClustersAction =
  ListClustersAction |
  SearchClustersAction |
  {type: 'FLOW_CATCH_ALL'};


export function searchClusters(search: string): SearchClustersAction {
  return {
    type: CLUSTERS_SEARCH,
    search: search,
  };
}
