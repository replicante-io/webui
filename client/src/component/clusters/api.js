'use strict';
//@flow

import type { ClusterItem } from '../dashboard/action';


export function searchClusters(search: string): Promise<Array<ClusterItem>> {
  let url = '/webui/clusters/search';
  let options = {
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({search: search}),
    method: 'POST',
  };
  return fetch(url, options).then((response) => {
    return response.json().then((body) => {
      if (!response.ok) {
        throw Error('Fetch error: ' + body.error);
      }
      return body;
    });
  });
}
