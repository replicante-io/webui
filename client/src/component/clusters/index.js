'use strict';
//@flow

import React from 'react';
import { connect } from 'react-redux';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/free-solid-svg-icons';

import ClusterRow from '../ClusterRow';

import { CLUSTERS_SEARCH } from './action';
import { searchClusters } from './action';

import type { ClusterMeta } from '../dashboard/action';
import type { ClustersStore } from './store';

// Re-export actions and store values.
export { defaultState, reducer } from './store';
export { saga } from './saga';
export type { ClustersStore } from './store';


const SEARCH_TIMEOUT = 500;  // ms


type Props = {
  clusters: Array<ClusterMeta>,
  dispatch: any,
  search: string,
};

type State = {
  searchTimeoutHandle: ?TimeoutID,
};


export class InnerClusters extends React.Component<Props, State> {
  input: any;

  constructor() {
    super();
    this.state = {
      searchTimeoutHandle: null,
    };
  };

  componentDidMount() {
    if (!this.props.clusters.length && !this.props.search) {
      this.props.dispatch(searchClusters(''));
    }
  }

  handleChange(e: SyntheticInputEvent<HTMLInputElement>) {
    let search = e.currentTarget.value;
    if (this.state.searchTimeoutHandle) {
      clearTimeout(this.state.searchTimeoutHandle);
    }
    let timeoutHandle = setTimeout(() => {
      this.props.dispatch(searchClusters(search));
      this.setState({searchTimeoutHandle: null});
    }, SEARCH_TIMEOUT);
    this.setState({searchTimeoutHandle: timeoutHandle});
  };

  renderList() {
    let rows = this.props.clusters.map((cluster) => {
      return <ClusterRow key={cluster.cluster_id} {...cluster} />;
    });
    return (
      <div className="card-body text-center">
        {rows}
      </div>
    );
  }

  renderNoResults() {
    return (
      <div className="card-body text-center">
        <p className="card-text large-icon"><FontAwesomeIcon icon={faSearch} /></p>
        <p className="card-text">
          No clusters were found.<br />
          Try searching something else.
        </p>
      </div>
    );
  };

  render() {
    const body = this.props.clusters.length ? this.renderList() : this.renderNoResults();
    return (
      <div className="row">
        <div className="col">
          <div className="card">
            <div className="card-header">
              Clusters
              <input type="text" className="cluster-search"
                  placeholder="Search ..."
                  defaultValue={this.props.search}
                  onChange={this.handleChange.bind(this)}
                  ref={(node) => this.input = node}
              />
            </div>
            {body}
          </div>
        </div>
      </div>
    );
  };
}


/** Map the redux state to properties passed to Button. */
type PartialState = {
  clusters: ClustersStore,
}
export function mapStateToProps(state: PartialState) {
  return {
    clusters: state.clusters.clusters,
    search: state.clusters.search,
  };
}


const Clusters = connect(mapStateToProps, null)(InnerClusters);
export default Clusters;
