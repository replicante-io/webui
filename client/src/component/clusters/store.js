'use strict';
//@flow

import { CLUSTERS_SEARCH } from './action';
import type { ClustersAction } from './action';
import type { SearchClustersAction } from './action';


export type ClustersStore = {
  search: string,
};


export const defaultState: ClustersStore = {
  search: '',
};


export function reducer(state: ClustersStore = defaultState, action: ClustersAction) {
  switch (action.type) {
    case CLUSTERS_SEARCH: return searchClusters(state, action);

    default:
      return state;
  }
}


function searchClusters(state: ClustersStore, action: SearchClustersAction): ClustersStore {
  let newState: ClustersStore = {...state};
  newState.search = action.search;
  return newState;
}
