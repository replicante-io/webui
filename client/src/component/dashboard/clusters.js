'use strict';
//@flow

import { Component } from 'react';
import React from 'react';

import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChartArea } from '@fortawesome/free-solid-svg-icons';

import { FetchButton } from '../datafetch';
import { fetchData } from '../datafetch';

import ClusterKinds from '../ClusterKinds';
import ClusterRow from '../ClusterRow';

import { storeClusters } from './action';

import type { ClusterMeta } from './action';
import type { DashboardStore } from './store';


const FETCH_ID: string = "top-clusters";


export function fetchClusters(dispatch: any): () => Promise<void> {
  return () => {
    return fetch('/webui/clusters/top').then((response) => {
      return response.json().then((body) => {
        if (!response.ok) {
          throw Error('Fetch error: ' + body.error);
        }
        dispatch(storeClusters(body));
      });
    });
  };
}


type Props = {
  clusters: Array<ClusterMeta>,
  dispatch: any,
}

export class InnerClusters extends React.Component<Props> {
  componentDidMount() {
    if (!this.props.clusters.length) {
      this.props.dispatch(fetchData(FETCH_ID, fetchClusters(this.props.dispatch)));
    }
  }

  renderClusters() {
    let rows = this.props.clusters.map((cluster) => {
      return <ClusterRow key={cluster.name} {...cluster} />;
    });
    return (
      <div className="card-body text-center">
        <div className="card-text">
          {rows}
        </div>
      </div>
    );
  }

  renderNoData() {
    return (
      <div className="card-body text-center">
        <p className="card-text large-icon"><FontAwesomeIcon icon={faChartArea} /></p>
        <p className="card-text">No data available</p>
      </div>
    );
  }

  render() {
    const body = this.props.clusters.length ? this.renderClusters() : this.renderNoData();
    return (
      <div className="row">
        <div className="col">
          <div className="card">
            <div className="card-header">
              Largest clusters
              <FetchButton id={FETCH_ID} request={fetchClusters(this.props.dispatch)} />
            </div>
            {body}
          </div>
        </div>
      </div>
    );
  }
}


/** Map the redux state to properties passed to Button. */
type PartialState = {
  dashboard: DashboardStore,
}
export function mapStateToProps(state: PartialState) {
  return {
    clusters: state.dashboard.clusters,
  };
}


const Clusters = connect(mapStateToProps, null)(InnerClusters);
export default Clusters;
