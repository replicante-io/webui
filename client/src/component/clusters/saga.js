'use strict';
//@flow

import { call, put } from 'redux-saga/effects';
import { takeEvery } from 'redux-saga/effects';

import { CLUSTERS_LIST } from './action';
import { CLUSTERS_SEARCH } from './action';
import { searchClusters as searchClustersApi } from './api';

import type { SearchClustersAction } from './action';


/**
 * Fetches data asyncronusly.
 *
 * The fetching process is actually perforemed by the
 * `action.start` function that must return a promise.
 */
export function* searchClusters(action: SearchClustersAction): any {
  let clusters = yield call(searchClustersApi, action.search);
  yield put({type: CLUSTERS_LIST, clusters: clusters});
}

/** Main saga for the datafetch component. */
export function* saga(): any {
  yield takeEvery(CLUSTERS_SEARCH, searchClusters);
}
