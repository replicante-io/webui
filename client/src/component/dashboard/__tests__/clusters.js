import fetch from 'jest-fetch-mock';
global.fetch = fetch;

import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import Clusters from '../clusters';
import { fetchClusters } from '../clusters';


const mockStore = configureStore([]);


describe('Dashboard', () => {
  describe('clusters', () => {

    test('renders with no clusters', () => {
      let clusters = [];
      let store = mockStore({
        datafetch: new Map(),
        dashboard: {clusters: clusters}
      });
      const tree = renderer.create(
        <Provider store={store}>
          <Clusters />
        </Provider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('renders with clusters', () => {
      let clusters = [{
        kinds: ['MongoDB'],
        name: 'test1',
        nodes: 1
      }, {
        kinds: ['Kafka'],
        name: 'test2',
        nodes: 4
      }];
      let store = mockStore({
        datafetch: new Map(),
        dashboard: {clusters: clusters}
      });
      const tree = renderer.create(
        <BrowserRouter>
          <Provider store={store}>
            <Clusters />
          </Provider>
        </BrowserRouter>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    describe('fetchClusters', () => {
      beforeEach(() => {
        fetch.resetMocks()
      });

      test('fulfill', () => {
        let fetcher = fetchClusters(jest.fn());
        fetch.mockResponse(JSON.stringify([{
          kinds: ['MongoDB'],
          name: 'test1',
          nodes: 1
        }, {
          kinds: ['Kafka'],
          name: 'test2',
          nodes: 4
        }]));
        return fetcher();
      });

      test('reject', () => {
        let error = Error('FAILED');
        let fetcher = fetchClusters(jest.fn());
        fetch.mockReject(error);
        return fetcher().then(
          () => { throw Error('Expected error') },
          (err) => { expect(err).toBe(error) }
        );
      });

      test('reject not 2xx', () => {
        let fetcher = fetchClusters(jest.fn());
        fetch.mockResponse('{}', {status: 500});
        return fetcher().then(
          () => { throw Error('Expected error') },
          () => { }
        );
      });
    });

  });
});
