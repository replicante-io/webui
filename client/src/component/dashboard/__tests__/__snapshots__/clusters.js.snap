// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard clusters renders with clusters 1`] = `
<div
  className="row"
>
  <div
    className="col"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        Largest clusters
        <button
          className="btn btn-outline-secondary fetch-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-redo-alt fa-w-16 "
            data-icon="redo-alt"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256.455 8c66.269.119 126.437 26.233 170.859 68.685l35.715-35.715C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.75c-30.864-28.899-70.801-44.907-113.23-45.273-92.398-.798-170.283 73.977-169.484 169.442C88.764 348.009 162.184 424 256 424c41.127 0 79.997-14.678 110.629-41.556 4.743-4.161 11.906-3.908 16.368.553l39.662 39.662c4.872 4.872 4.631 12.815-.482 17.433C378.202 479.813 319.926 504 256 504 119.034 504 8.001 392.967 8 256.002 7.999 119.193 119.646 7.755 256.455 8z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
      <div
        className="card-body text-center"
      >
        <div
          className="card-text"
        >
          <div
            className="row cluster-item"
          >
            <div
              className="col"
            >
              <p>
                <a
                  href="/clusters/test1"
                  onClick={[Function]}
                >
                  test1
                </a>
              </p>
              <small>
                Name
              </small>
            </div>
            <div
              className="col"
            >
              <p>
                1
              </p>
              <small>
                Nodes
              </small>
            </div>
            <div
              className="col"
            >
              <p>
                <span
                  className="db-icon mongodb"
                  title="MongoDB"
                />
              </p>
              <small>
                MongoDB
              </small>
            </div>
          </div>
          <div
            className="row cluster-item"
          >
            <div
              className="col"
            >
              <p>
                <a
                  href="/clusters/test2"
                  onClick={[Function]}
                >
                  test2
                </a>
              </p>
              <small>
                Name
              </small>
            </div>
            <div
              className="col"
            >
              <p>
                4
              </p>
              <small>
                Nodes
              </small>
            </div>
            <div
              className="col"
            >
              <p>
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-database fa-w-14 "
                  data-icon="database"
                  data-prefix="fas"
                  role="img"
                  style={Object {}}
                  title="Kafka"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M448 73.143v45.714C448 159.143 347.667 192 224 192S0 159.143 0 118.857V73.143C0 32.857 100.333 0 224 0s224 32.857 224 73.143zM448 176v102.857C448 319.143 347.667 352 224 352S0 319.143 0 278.857V176c48.125 33.143 136.208 48.572 224 48.572S399.874 209.143 448 176zm0 160v102.857C448 479.143 347.667 512 224 512S0 479.143 0 438.857V336c48.125 33.143 136.208 48.572 224 48.572S399.874 369.143 448 336z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </p>
              <small>
                Kafka
              </small>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dashboard clusters renders with no clusters 1`] = `
<div
  className="row"
>
  <div
    className="col"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        Largest clusters
        <button
          className="btn btn-outline-secondary fetch-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-redo-alt fa-w-16 "
            data-icon="redo-alt"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256.455 8c66.269.119 126.437 26.233 170.859 68.685l35.715-35.715C478.149 25.851 504 36.559 504 57.941V192c0 13.255-10.745 24-24 24H345.941c-21.382 0-32.09-25.851-16.971-40.971l41.75-41.75c-30.864-28.899-70.801-44.907-113.23-45.273-92.398-.798-170.283 73.977-169.484 169.442C88.764 348.009 162.184 424 256 424c41.127 0 79.997-14.678 110.629-41.556 4.743-4.161 11.906-3.908 16.368.553l39.662 39.662c4.872 4.872 4.631 12.815-.482 17.433C378.202 479.813 319.926 504 256 504 119.034 504 8.001 392.967 8 256.002 7.999 119.193 119.646 7.755 256.455 8z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
      <div
        className="card-body text-center"
      >
        <p
          className="card-text large-icon"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-chart-area fa-w-16 "
            data-icon="chart-area"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M500 384c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v308h436zM372.7 159.5L288 216l-85.3-113.7c-5.1-6.8-15.5-6.3-19.9 1L96 248v104h384l-89.9-187.8c-3.2-6.5-11.4-8.7-17.4-4.7z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </p>
        <p
          className="card-text"
        >
          No data available
        </p>
      </div>
    </div>
  </div>
</div>
`;
